src/realm.js

```diff
@@ -1295,16 +1310,49 @@ export class Realm {
 
   // Record the current value of binding in this.modifiedBindings unless
   // there is already an entry for binding.
-  recordModifiedBinding(binding: Binding): Binding {
+  recordModifiedBinding(binding: Binding, value?: Value): Binding {
     if (binding.environment.isReadOnly) {
       // This only happens during speculative execution and is reported elsewhere
       throw new FatalError("Trying to modify a binding in read-only realm");
     }
-    if (this.modifiedBindings !== undefined && !this.modifiedBindings.has(binding))
+
+    const isEnvSafe = root => {
+      let context = this.getRunningContext();
```

Comment:
I don't think the currently running context matters. What matters is where the binding was defined. You need to pass in the environment from the binding here.

I'll try to think of a test case but pretty sure this is wrong.

Reply:
A test case would be awesome :)

Reply:
```js
var x = 0;
function foo() {
  x++; // not safe
}
function bar() {
  foo();
}
__pure(bar);
```

```js
var x = 0;
function foo() {
  function foo2() {
    x++; // not safe
  }
  foo2();
}
function bar() {
  foo();
}
__pure(bar);
```

```js
function foo() {
  var x = 0;
  function foo2() {
    x++; // safe
  }
  foo2();
}
function bar() {
  foo();
}
__pure(bar);
```

```js
function foo() {
  var x = 0;
  function foo2() {
    function foo3() {
      x++; // safe
    }
    foo3();
  }
  foo2();
}
function bar() {
  foo();
}
__pure(bar);
```

Reply:
All these tests pass with the current code.